ext/debug/ctf/README.rst

```diff
@@ -0,0 +1,204 @@
+
+In fact, IO varies greatly from system to system.  Therefore, it is
+instructive to create a taxonomy for IO types when we must ensure the
+interface between CTF-middle and CTF-bottom is generic and efficient
+enough to model these. See the *IO taxonomy* section below.
+
+
+A generic interface
+-------------------
+
+In CTF, an event is serialized to a packet containing one or more fields
+and as seen from *IO taxonomy* section below, a bottom layer:
+
+- may perform action at transaction-start (e.g. mutex-lock),
+- may process each field in some way (e.g. sync-push emit, concat, enqueue to thread-bound FIFO),
+- may perform action at transaction-stop (e.g. mutex-release, emit of concat buffer).
+
+This is possible when the bottom-layer implements the following macros:
+
+- ``CTF_BOTTOM_LOCK``:   No-op or how to lock the IO transaction.
+- ``CTF_BOTTOM_UNLOCK``: No-op or how to release the IO transaction.
+- ``CTF_BOTTOM_FIELDS``: Var-args of fields. May process each field with ``MAP``.
+- ``CTF_BOTTOM_TIMESTAMPED_INTERNALLY``: Tells where timestamping is done.
+
+These macros along with inline functions of the middle-layer can yield
+very low-overhead tracing infrastructure.
+
+
+CTF middle-layer example
+------------------------
+
+The CTF_EVENT macro will serialize each argument to a field::
+
+  /* Example for illustration */
+  static inline void ctf_middle_foo(u32_t thread_id, ctf_bounded_string_t name)
+  {
+    CTF_EVENT(
+      CTF_LITERAL(u8_t, 42),
+      thread_id,
+      name,
+      "hello, I was emitted from function: ",
+      __func__  /* __func__ is standard since C99 */
+    );
+  }
+
+How to serialize and emit fields as well as handling alignment, can be done
+internally and statically at compile-time in the bottom-layer.
+
+
+How to activate?
+----------------
+
+Make sure ``CONFIG_CTF=y`` and ``CONFIG_CTF_BOTTOM_POSIX=y`` is set.
+
+
+
+What is TraceCompass?
+---------------------
+
+TraceCompass is an open source tool that visualizes CTF events such
+as thread scheduling and interrupts, and is helpful to find unintended
+interactions and resource conflicts on complex systems.
+
+See also the presentation by Ericsson,
+`Advanced Trouble-shooting Of Real-time Systems
+<https://wiki.eclipse.org/images/0/0e/TechTalkOnlineDemoFeb2017_v1.pdf>`_.
+
+
+Future LTTng inspiration
+------------------------
+
+Currently, the middle-layer provided here is quite simple and bare-bones,
+and needlessly copied from Zephyr's Segger SystemView debug module.
+
+For an OS like Zephyr, it would make sense to draw inspiration from
+Linux' LTTng and change the middle-layer to serialize to the same format.
+Doing this would enable direct reuse of TraceCompass' canned analyses
+for Linux.  Alternatively, LTTng-analyses in TraceCompass could be
+customized to Zephyr.  It is ongoing work to enable TraceCompass
+visibility of Zephyr in a target-agnostic and open source way.
+
+
+IO taxonomy
+-----------
+
+  - Atomic Push/Produce/Write/Enqueue:
```

Comment:
This whole section is unnecessarily indented by two extra spaces that would negatively affect the generated HTML